### PR TITLE
Add metrics to the Prometheus API

### PR DESCRIPTION
**Pull request by shkwsk**
Hello.

This patch adds latency and availability metrics for Predict method in gRPC API to the Prometheus API.
The metrics are measured for the entire process and for each model.

The added metric items are as follows.
* /tensorflow/serving/all_request_count: the total number of requests in requests processing
* /tensorflow/serving/all_request_fail_count: number of failed requests in the entire request process
* /tensorflow/serving/all_request_latency_usec: total request processing latency
* /tensorflow/serving/all_request_latency_histogram_usec: a histogram of the total latency of the request processing
* /tensorflow/serving/model_request_count: number of requests by model
* /tensorflow/serving/model_request_fail_count: number of failed requests by model
* /tensorflow/serving/model_request_latency_usec: latency by model
* /tensorflow/serving/model_request_latency_histogram_usec: latency histogram by model

### CODE REVIEW

**Comment by qisikai** [comment]
Great Job！

**Review by qisikai — DISMISSED** [review]

**Comment by shkwsk** [comment]
It's been two weeks now, could you please confirm this?

**Comment by Windfarer** [comment]
any update about this PR? will it be merged? @qisikai @netfs

**Comment by shkwsk** [comment]
@netfs Please check and merge this PR?

**Comment by netfs** [comment]
please do look at c317582981cfc that adds runtime (latency) metrics for predict/regress/classify methods (and these are exported as prometheus metrics, if you enable prometheus export).

the above commit does not include latency stats for rpc/http handling, and thats where your PR would be useful.

**Comment by shkwsk** [comment]
@netfs 
Yes.
I want to measure the latency stats for rpc/http handling.
And I also want to measure the availability of the model.

This PR allows to measure availability and latency that not includes in c317582.

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 1]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+
+auto* model_latency_histogram = monitoring::Sampler<1>::New(
+    {"/tensorflow/serving/model_request_latency_histogram_usec",							    
+     "The total time spent on executing graphs in microseconds.", "model"},
+    // It would be nice to be able to set the parameters flexibly.
+    monitoring::Buckets::Explicit({
+	1000, 2000, 3000, 4000, 5000, 7000, 9000, 11000, 13000, 15000,
+	17000, 19000, 21000, 24000, 27000, 30000, 33000, 35000, 38000}));
+
+// All processing metrics
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:71]: metric aggregator/collector systems can sum things up based on per-model counters (above). no need to add `all_` counters.

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 2]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
     "/tensorflow/serving/request_example_count_total",
     "The total number of tensorflow.Examples.", "model");
 
+// Metrics by model
+auto* model_request_count_total = monitoring::Counter<1>::New(
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:51]: change this to Counter<2> to collect per-model+status count. so you can collect success and errors together.

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 3]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+    "/tensorflow/serving/model_request_count",
+    "The total number of requests.", "model");
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:53]: s/model/model_name/

Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:53]: add additional dimension "status" to capture status, you can use the string version of error code enums to populate these.

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 4]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+
+auto* model_request_fail_count_total = monitoring::Counter<1>::New(
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:55]: remove this if you are tracking per-model+error-code counts in previous counter.

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 5]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+    "/tensorflow/serving/model_request_fail_count",
+    "The total number of faled requests.", "model");
+
+auto* model_latency_total = monitoring::Counter<1>::New(
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:59]: why do we need this? can't aggregator/collectors sum this up from the latency distribution (next counter)?

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 6]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+    "/tensorflow/serving/model_request_latency_usec",
+    "The total time spent on executing graphs in microseconds.", "model");
+
+auto* model_latency_histogram = monitoring::Sampler<1>::New(
+    {"/tensorflow/serving/model_request_latency_histogram_usec",							    
+     "The total time spent on executing graphs in microseconds.", "model"},
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:65]: s/model/model_name/

File: tensorflow_serving/servables/tensorflow/util.cc [hunk 7]
```diff
@@ -47,6 +47,48 @@ auto* example_count_total = monitoring::Counter<1>::New(
...
+    // It would be nice to be able to set the parameters flexibly.
+    monitoring::Buckets::Explicit({
```
Comment (netfs) [inline tensorflow_serving/servables/tensorflow/util.cc:67]: can these buckets be used:

https://github.com/tensorflow/serving/blob/e206489e91a6c20b5841bf954e428870bc7b5079/tensorflow_serving/servables/tensorflow/util.cc#L65

Comment (shkwsk) [reply tensorflow_serving/servables/tensorflow/util.cc]: I have pull-requested earlier than that commit.
Oh well, I'll fix it to use this one.

**Comment by shkwsk** [comment]
@netfs I fixed to your reviews.

**Review by netfs — COMMENTED** [review]
thanks for the recent changes. few more suggestions.

File: tensorflow_serving/model_servers/http_server.cc [hunk 8]
```diff
@@ -169,6 +173,10 @@ class RestApiRequestDispatcher {
               << " " << req->uri_path() << " Error: " << status.ToString();
     }
     req->ReplyWithStatus(http_status);
+
+    RecordRuntimeLatency(model_name, /*api=*/method, /*runtime=*/"HTTP",
```
[the remark this answers is not in the document]
Comment (shkwsk) [reply tensorflow_serving/model_servers/http_server.cc]: OK, it's enough to distinguish between grpc and rest, so I named the dimension "entrypoint".

**Comment by shkwsk** [comment]
@netfs I fixed.

File: tensorflow_serving/model_servers/prediction_service_impl.cc [hunk 9]
```diff
@@ -51,18 +52,24 @@ thread::ThreadPoolOptions GetThreadPoolOptions(
...
-  const ::grpc::Status status =
-      ToGRPCStatus(predictor_->Predict(run_options, core_, *request, response));
+  const ::tensorflow::Status tf_status = predictor_->Predict(
+      run_options, core_, *request, response);
+  const ::grpc::Status status = ToGRPCStatus(tf_status);
 
   if (!status.ok()) {
     VLOG(1) << "Predict failed: " << status.error_message();
   }
+  RecordRequestLatency(request->model_spec().name(), /*api=*/"Predict",
```
Comment (netfs) [inline tensorflow_serving/model_servers/prediction_service_impl.cc:69]: (here and elsewhere), we do not want to record latency for requests that error out. no?

**Review by netfs — COMMENTED** [review]
also please fix broken test:

https://source.cloud.google.com/results/invocations/efea5490-a7cc-4556-a1ac-115b7c25a0d9/targets

File: tensorflow_serving/model_servers/prediction_service_impl.cc (hunk 9 above)
Comment (shkwsk) [reply tensorflow_serving/model_servers/prediction_service_impl.cc:69]: OK, I try to add status label in request_latency.

File: tensorflow_serving/model_servers/http_server.cc [hunk 10]
```diff
@@ -176,7 +176,7 @@ class RestApiRequestDispatcher {
     req->ReplyWithStatus(http_status);
 
     RecordRequestLatency(model_name, /*api=*/method, /*entrypoint=*/"REST",
-                         Env::Default()->NowMicros() - start);
+                         status, Env::Default()->NowMicros() - start);
```
Comment (netfs) [inline tensorflow_serving/model_servers/http_server.cc:179]: maybe i should have clarified. i think we only need to track of latency of **successful** requests. for errored out requests, i do not see it being relevant. you want to remove "status" as a dimension and only increment/record latency for successful requests at call site.

File: tensorflow_serving/model_servers/http_server.cc [hunk 11]
```diff
@@ -169,6 +174,12 @@ class RestApiRequestDispatcher {
               << " " << req->uri_path() << " Error: " << status.ToString();
     }
     req->ReplyWithStatus(http_status);
+
```
Comment (netfs) [inline tensorflow_serving/model_servers/http_server.cc:177]: will this be better/readable:

```
if (http_status == net_http::HTTPStatusCode::OK) {
   RecordRequestLatency(model_name, /*api=*/method, /*entrypoint=*/"REST",
                                           Env::Default()->NowMicros() - start);
} else {
   VLOG(1) << "Error Processing HTTP/REST request: " << req->http_method()
                  << " " << req->uri_path() << " Error: " << status.ToString();
}
RecordModelRequestCount(model_name, status);
req->ReplyWithStatus(http_status);
```

so we do not time any socket write delay, the way we do not account for the read side when we get the request, and start our timing when we being processing the request above. this makes it symmetric.

**Comment by netfs** [comment]
Will merge shortly.

**Comment by shkwsk** [comment]
Sorry, I fixed test.

**Review by netfs — APPROVED** [review]

**Pull request merged**